### PR TITLE
koji_promote: don't encode docker logs and final openshift logs if logs are bytestream

### PR DESCRIPTION
**Pull request by vrutkovs**
Signed-off-by: Vadim Rutkovsky <vrutkovs@redhat.com>

### CODE REVIEW

File: rel-eng/packages/atomic-reactor [hunk 1]
```diff
@@ -1 +1 @@
-1.6.23.1-1 ./
+1.6.23-1 ./
```
Comment (rcerven) [inline rel-eng/packages/atomic-reactor]: latest release version is 1.6.23.2

so not sure if there should be future version or current version

Comment (vrutkovs) [inline rel-eng/packages/atomic-reactor]: Right, that doesn't belong here. Removed

**Comment by twaugh** [comment]
Don't forget https://github.com/projectatomic/atomic-reactor/blob/master/tests/plugins/test_orchestrate_build.py#L139.

File: atomic_reactor/plugins/exit_koji_promote.py [hunk 2]
```diff
@@ -354,7 +354,7 @@ def get_logs(self):
         docker_logs = NamedTemporaryFile(prefix="docker-%s" % self.build_id,
                                          suffix=".log",
                                          mode='wb')
-        docker_logs.write("\n".join(self.workflow.build_result.logs).encode('UTF-8'))
+        docker_logs.write(b"\n".join(self.workflow.build_result.logs))
```
Comment (twaugh) [inline atomic_reactor/plugins/exit_koji_promote.py]: This value comes from atomic-reactor itself and wouldn't have changed type I think.

File: atomic_reactor/plugins/exit_koji_promote.py [hunk 3]
```diff
@@ -345,7 +345,7 @@ def get_logs(self):
             logfile = NamedTemporaryFile(prefix=self.build_id,
                                          suffix=".log",
                                          mode='wb')
-            logfile.write(logs.encode('UTF-8'))
+            logfile.write(logs)
```
Comment (twaugh) [inline atomic_reactor/plugins/exit_koji_promote.py]: I believe this should be compatible with old and new versions of osbs-client.

It's just that we need this new version of atomic-reactor when using the new osbs-client in the builder image they are both in.

Comment (vrutkovs) [inline atomic_reactor/plugins/exit_koji_promote.py]: Hmm, right. Not sure how to do that correctly - use `six` find out the string type?

File: tests/plugins/test_koji_promote.py [hunk 4]
```diff
@@ -261,10 +261,10 @@ def mock_environment(tmpdir, session=None, name=None,
         setattr(workflow, 'exported_image_sequence', [{'path': fp.name}])
 
     if build_process_failed:
-        workflow.build_result = BuildResult(logs=["docker build log - \u2018 \u2017 \u2019 \n'"],
+        workflow.build_result = BuildResult(logs=[b"docker build log - \u2018 \u2017 \u2019 \n'"],
```
Comment (twaugh) [inline tests/plugins/test_koji_promote.py]: I think this should not change.

File: tests/plugins/test_koji_promote.py [hunk 5]
```diff
@@ -261,10 +261,10 @@ def mock_environment(tmpdir, session=None, name=None,
...
                                             fail_reason="not built")
     else:
-        workflow.build_result = BuildResult(logs=["docker build log - \u2018 \u2017 \u2019 \n'"],
+        workflow.build_result = BuildResult(logs=[b"docker build log - \u2018 \u2017 \u2019 \n'"],
```
Comment (twaugh) [inline tests/plugins/test_koji_promote.py]: I think this should not change.

File: atomic_reactor/plugins/exit_koji_promote.py (hunk 3 above)
Comment (twaugh) [inline atomic_reactor/plugins/exit_koji_promote.py]: I *think* it might be OK as it is.

File: tests/plugins/test_koji_promote.py (hunk 4 above)
Comment (vrutkovs) [inline tests/plugins/test_koji_promote.py]: Ah, correct, these are logs straight from container

File: atomic_reactor/plugins/exit_koji_promote.py [hunk 6]
```diff
@@ -345,7 +345,11 @@ def get_logs(self):
             logfile = NamedTemporaryFile(prefix=self.build_id,
                                          suffix=".log",
                                          mode='wb')
-            logfile.write(logs.encode('UTF-8'))
+            try:
+                logfile.write(logs)
+            except UnicodeEncodeError:
```
Comment (twaugh) [inline atomic_reactor/plugins/exit_koji_promote.py]: No, `TypeError`.

Comment (twaugh) [inline atomic_reactor/plugins/exit_koji_promote.py]: Or do different versions of Python behave differently with this? I hope not.

Comment (vrutkovs) [inline atomic_reactor/plugins/exit_koji_promote.py]: It doesn't throw `TypeError` :( Not sure why though, unit test looks correct to me

Comment (twaugh) [inline atomic_reactor/plugins/exit_koji_promote.py]: Python 3 throws TypeError:
```
>>> open('/tmp/foo', 'wb').write('ä')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: a bytes-like object is required, not 'str'
```

So maybe catch both.

Comment (vrutkovs) [inline atomic_reactor/plugins/exit_koji_promote.py]: Oh, I see. So it throws `TypeError` in py3, as expected.

The problem that it currently throws `UnicodeEncodeError` in py2 in this case

Comment (twaugh) [inline atomic_reactor/plugins/exit_koji_promote.py]: We can just `except (TypeError, UnicodeEncodeError):` our way around it I expect.

Comment (vrutkovs) [inline atomic_reactor/plugins/exit_koji_promote.py]: Right, that worked

**Comment by vrutkovs** [comment]
Integration tests seem to pass with osbs-clients PR and this PR.

Not sure how to verify py3 case though - I'll try to revive fedora-based osbs-box

**Comment by lcarva** [comment]
> Not sure how to verify py3 case though

This plugin requires koji library which is only python 2, safe to ignore it?

**Comment by twaugh** [comment]
I'd say so, yes.

**Comment by vrutkovs** [comment]
>This plugin requires koji library which is only python 2, safe to ignore it?

Ah, yes. 

Koji module would be py3-compatible shortly though, so more fixes might be required

**Pull request merged**